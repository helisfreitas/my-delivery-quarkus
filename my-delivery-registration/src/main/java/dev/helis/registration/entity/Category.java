package dev.helis.registration.entity;

public enum Category {

    STARTER, MAIN_COURSE, DESSERT;
}
