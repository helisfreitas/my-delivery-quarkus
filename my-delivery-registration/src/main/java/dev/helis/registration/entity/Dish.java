package dev.helis.registration.entity;

import java.math.BigDecimal;
import java.net.URL;

import dev.helis.registration.validation.OnlyCharacterAndPunctuation;

import io.quarkus.hibernate.orm.panache.PanacheEntityBase;
import io.smallrye.common.constraint.NotNull;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.ManyToOne;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Positive;

@Entity
public class Dish extends PanacheEntityBase {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    public Long id;

    @NotBlank
    @OnlyCharacterAndPunctuation
    public String name;

    @OnlyCharacterAndPunctuation
    public String description;

    @NotNull
    @Positive
    public BigDecimal price;

    public URL image;

    @Enumerated(EnumType.STRING)
    public Category category;

    public Boolean isAvailable = Boolean.TRUE;

    @ManyToOne
    public Restaurant restaurant;

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + ((name == null) ? 0 : name.hashCode());
        result = prime * result + ((category == null) ? 0 : category.hashCode());
        result = prime * result + ((restaurant == null) ? 0 : restaurant.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;
        Dish other = (Dish) obj;
        if (name == null) {
            if (other.name != null)
                return false;
        } else if (!name.equals(other.name))
            return false;
        if (category != other.category)
            return false;
        if (restaurant == null) {
            if (other.restaurant != null)
                return false;
        } else if (!restaurant.equals(other.restaurant))
            return false;
        return true;
    }

    @Override
    public String toString() {
        return "Dish [name=" + name + ", price=" + price + ", category=" + category + ", restaurant=" + restaurant
                + "]";
    }    

}
