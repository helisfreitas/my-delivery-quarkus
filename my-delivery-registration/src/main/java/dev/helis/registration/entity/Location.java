package dev.helis.registration.entity;

import io.quarkus.hibernate.orm.panache.PanacheEntityBase;
import io.smallrye.common.constraint.NotNull;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;

@Entity
public class Location extends PanacheEntityBase {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    Long id;

    @NotNull
    public double longitude;

    @NotNull
    public double latitude;

    public Location() {
        // Default constructor
    }

    public Location(String longitude, String latitude) {
        this.longitude = convertDMSToDecimal(longitude);
        this.latitude = convertDMSToDecimal(latitude);
    }

    public Location(double longitude, double latitude) {
        this.longitude = longitude;
        this.latitude = latitude;
    }

    public void setLatitude(String latitude) {
        this.latitude = convertDMSToDecimal(latitude);
    }

    public void setLongitude(String longitude) {
        this.longitude = convertDMSToDecimal(longitude);
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        long temp;
        temp = Double.doubleToLongBits(longitude);
        result = prime * result + (int) (temp ^ (temp >>> 32));
        temp = Double.doubleToLongBits(latitude);
        result = prime * result + (int) (temp ^ (temp >>> 32));
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null || getClass() != obj.getClass())
            return false;
        Location other = (Location) obj;
        return Double.compare(this.longitude, other.longitude) == 0 && Double.compare(this.latitude, other.latitude) == 0;
    }

    @Override
    public String toString() {
        return "Location [longitude=" + longitude + ", latitude=" + latitude + "]";
    }
 
    public String convertLatitudeToDMS() {
        return convertToDMS(latitude, true);
    }

    public String convertLongitudeToDMS() {
        return convertToDMS(longitude, false);
    }
     
    private String convertToDMS(double coordinate, boolean isLatitude) {
        // Obtendo o símbolo de direção
        String direction;
        if (isLatitude) {
            direction = (coordinate >= 0) ? "N" : "S";
        } else {
            direction = (coordinate >= 0) ? "E" : "W";
        }

        // Convertendo para valor absoluto (tirar o negativo para o cálculo)
        coordinate = Math.abs(coordinate);

        // Calculando graus, minutos e segundos
        int graus = (int) coordinate;
        double decimalMinutes = (coordinate - graus) * 60;
        int minutos = (int) decimalMinutes;
        double segundos = (decimalMinutes - minutos) * 60;

        // Formatando o resultado
        return String.format("%d°%d'%4.1f\"%s", graus, minutos, segundos, direction);
    }

    public double convertDMSToDecimal(String dms) {
        // Identificando a direção
        char direction = dms.charAt(dms.length() - 1);
        boolean isNegative = (direction == 'S' || direction == 'W');

        // Remover o símbolo de direção para fazer o cálculo
        dms = dms.substring(0, dms.length() - 1);

        // Dividindo a string pelos componentes de graus, minutos e segundos
        String[] dmsParts = dms.split("[°'\"]");

        // Extraindo graus, minutos e segundos
        int degrees = Integer.parseInt(dmsParts[0].trim());
        int minutes = Integer.parseInt(dmsParts[1].trim());
        double seconds = Double.parseDouble(dmsParts[2].trim());

        // Convertendo para decimal
        double decimal = degrees + (minutes / 60.0) + (seconds / 3600.0);

        // Aplicando o sinal negativo, se necessário
        return isNegative ? -decimal : decimal;
    }


}
