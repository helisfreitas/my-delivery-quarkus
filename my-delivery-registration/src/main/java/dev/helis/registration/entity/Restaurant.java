package dev.helis.registration.entity;

import java.time.LocalDate;

import dev.helis.registration.validation.OnlyCharacterAndPunctuation;

import io.quarkus.hibernate.orm.panache.PanacheEntityBase;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.ManyToOne;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.PastOrPresent;

@Entity
public class Restaurant extends PanacheEntityBase {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    public Long id;

    @NotBlank
    public String owner;

    @NotBlank
    @OnlyCharacterAndPunctuation
    public String name;

    @ManyToOne(cascade = CascadeType.ALL)
    public Location location;

    @PastOrPresent
    public LocalDate creationDate;

    @PastOrPresent
    public LocalDate lastUpdate;

    public Restaurant() {

        // Default constructor
    }

    public Restaurant(String name, String owner, Location location) {
        this.owner = owner;
        this.name = name;
        this.location = location;
        this.creationDate = LocalDate.now();
        this.lastUpdate = LocalDate.now();
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + ((owner == null) ? 0 : owner.hashCode());
        result = prime * result + ((name == null) ? 0 : name.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;
        Restaurant other = (Restaurant) obj;
        if (owner == null) {
            if (other.owner != null)
                return false;
        } else if (!owner.equals(other.owner))
            return false;
        if (name == null) {
            if (other.name != null)
                return false;
        } else if (!name.equals(other.name))
            return false;
        return true;
    }

    @Override
    public String toString() {
        return "Restaurant [owner=" + owner + ", name=" + name + "]";
    }

}
