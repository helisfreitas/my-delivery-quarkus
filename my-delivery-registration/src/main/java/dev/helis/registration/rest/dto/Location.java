package dev.helis.registration.rest.dto;

import java.io.Serializable;

import jakarta.validation.constraints.Size;

public class Location implements Serializable {

    private static final long serialVersionUID = 1L;

    @Size(min = 11, max = 12)
    private String longitude;

    @Size(min = 11, max = 12)
    private String latitude;

    public Location() {
        // Default constructor
    }

    public Location(String longitude, String latitude) {
        this.longitude = longitude;
        this.latitude = latitude;
    }

    public String getLongitude() {
        return longitude;
    }

    public void setLongitude(String longitude) {
        this.longitude = longitude;
    }

    public String getLatitude() {
        return latitude;
    }

    public void setLatitude(String latitude) {
        this.latitude = latitude;
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + ((longitude == null) ? 0 : longitude.hashCode());
        result = prime * result + ((latitude == null) ? 0 : latitude.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;
        Location other = (Location) obj;
        if (longitude == null) {
            if (other.longitude != null)
                return false;
        } else if (!longitude.equals(other.longitude))
            return false;
        if (latitude == null) {
            if (other.latitude != null)
                return false;
        } else if (!latitude.equals(other.latitude))
            return false;
        return true;
    }

    @Override
    public String toString() {
        return "Location [longitude=" + longitude + ", latitude=" + latitude + "]";
    }

    
}
