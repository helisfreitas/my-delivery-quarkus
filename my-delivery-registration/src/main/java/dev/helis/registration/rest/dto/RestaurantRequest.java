package dev.helis.registration.rest.dto;

import dev.helis.registration.validation.OnlyCharacterAndPunctuation;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;

public class RestaurantRequest implements java.io.Serializable {

    private static final long serialVersionUID = 1L;

    @NotBlank
    @Size(min = 5, max = 100)
    private String owner;

    @NotBlank
    @Size(min = 5, max = 100)
    @OnlyCharacterAndPunctuation
    private String name;

    @NotNull  
    private Location location;

    public RestaurantRequest() {
        // Default constructor
    }

    public RestaurantRequest(@NotBlank String owner, @NotBlank String name, @NotNull Location location) {
        this.owner = owner;
        this.name = name;
        this.location = location;
    }

    public String getOwner() {
        return owner;
    }

    public void setOwner(String owner) {
        this.owner = owner;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Location getLocation() {
        return location;
    }

    public void setLocation(Location location) {
        this.location = location;
    }

    @Override
    public String toString() {
        return "RestaurantRequest [owner=" + owner + ", name=" + name + ", location=" + location + "]";
    }   

}
