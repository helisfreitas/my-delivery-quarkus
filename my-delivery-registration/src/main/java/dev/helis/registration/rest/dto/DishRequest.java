package dev.helis.registration.rest.dto;

import java.io.Serializable;
import java.math.BigDecimal;

import dev.helis.registration.entity.Category;
import dev.helis.registration.validation.OnlyCharacterAndPunctuation;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;

public class DishRequest implements Serializable {

    private static final long serialVersionUID = 1L;

    @NotBlank
    @OnlyCharacterAndPunctuation
    private String name;

    @OnlyCharacterAndPunctuation
    private String description;

    @NotNull
    @Positive
    private BigDecimal price;

    private String image;

    @NotNull
    private Category category;

    @NotNull
    private Boolean isAvailable;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public BigDecimal getPrice() {
        return price;
    }

    public void setPrice(BigDecimal price) {
        this.price = price;
    }

    public String getImage() {
        return image;
    }

    public void setImage(String image) {
        this.image = image;
    }

    public Category getCategory() {
        return category;
    }

    public void setCategory(Category category) {
        this.category = category;
    }

    public Boolean getIsAvailable() {
        return isAvailable;
    }

    public void setIsAvailable(Boolean isAvailable) {
        this.isAvailable = isAvailable;
    }

    @Override
    public String toString() {
        return "Dish [name=" + name + ", price=" + price + ", category=" + category
                + "]";
    }

}
