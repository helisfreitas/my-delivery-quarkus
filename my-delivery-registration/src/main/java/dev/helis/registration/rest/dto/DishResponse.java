package dev.helis.registration.rest.dto;

import java.io.Serializable;
import java.math.BigDecimal;
import java.net.URL;

import dev.helis.registration.entity.Category;


public class DishResponse implements Serializable {

    private static final long serialVersionUID = 1L;

    private Long id;

    private String name;

    private String description;

    private BigDecimal price;

    private URL image;

    private Category category;

    private Boolean isAvailable;

    private Long restaurant;
    

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public BigDecimal getPrice() {
        return price;
    }

    public void setPrice(BigDecimal price) {
        this.price = price;
    }

    public URL getImage() {
        return image;
    }

    public void setImage(URL image) {
        this.image = image;
    }

    public Category getCategory() {
        return category;
    }

    public void setCategory(Category category) {
        this.category = category;
    }

    public Boolean getIsAvailable() {
        return isAvailable;
    }

    public void setIsAvailable(Boolean isAvailable) {
        this.isAvailable = isAvailable;
    }

    public Long getRestaurant() {
        return restaurant;
    }

    public void setRestaurant(Long restaurant) {
        this.restaurant = restaurant;
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + ((name == null) ? 0 : name.hashCode());
        result = prime * result + ((category == null) ? 0 : category.hashCode());
        result = prime * result + ((restaurant == null) ? 0 : restaurant.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;
        DishResponse other = (DishResponse) obj;
        if (name == null) {
            if (other.name != null)
                return false;
        } else if (!name.equals(other.name))
            return false;
        if (category != other.category)
            return false;
        if (restaurant == null) {
            if (other.restaurant != null)
                return false;
        } else if (!restaurant.equals(other.restaurant))
            return false;
        return true;
    }

    @Override
    public String toString() {
        return "Dish [name=" + name + ", price=" + price + ", category=" + category + ", restaurant=" + restaurant
                + "]";
    }    

}
