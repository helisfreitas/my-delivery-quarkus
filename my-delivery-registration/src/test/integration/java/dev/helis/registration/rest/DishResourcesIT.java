package dev.helis.registration.rest;

import static io.restassured.RestAssured.given;

import java.math.BigDecimal;

import org.approvaltests.Approvals;
import org.approvaltests.JsonApprovals;
import org.junit.jupiter.api.Tag;

import com.github.database.rider.cdi.api.DBRider;
import com.github.database.rider.core.api.configuration.DBUnit;
import com.github.database.rider.core.api.configuration.Orthography;
import com.github.database.rider.core.api.dataset.DataSet;
import com.github.database.rider.core.api.dataset.ExpectedDataSet;

import dev.helis.helper.RegistrationTestLifecycleManager;
import dev.helis.helper.annotation.ErrorTest;
import dev.helis.helper.annotation.IntegrationRestTest;
import dev.helis.helper.annotation.SuccessTest;
import dev.helis.registration.entity.Category;
import dev.helis.registration.rest.dto.DishRequest;
import io.quarkus.test.common.QuarkusTestResource;
import io.quarkus.test.junit.QuarkusTest;
import jakarta.transaction.Transactional;

@QuarkusTest
@QuarkusTestResource(RegistrationTestLifecycleManager.class)
@DBRider
@DBUnit(caseInsensitiveStrategy = Orthography.LOWERCASE, alwaysCleanBefore = true)
@IntegrationRestTest
@Tag("dish-feature")
class DishResourcesIT {

    @SuccessTest
    @DataSet(value = "/DishResourcesIT/dishes-scenario-2.yml")
    @Transactional
    void shouldFindAllDishesFromRestaurant() {

        String response = given()
                .when().get(ResourcePaths.DISHES, 1)
                .then()
                .statusCode(200)
                .extract().asString();

        JsonApprovals.verifyJson(response);
    }

    @SuccessTest
    @DataSet(value = "/DishResourcesIT/dishes-scenario-2.yml")
    @Transactional
    void shouldFindByIdDishesFromRestaurant() {

        String response = given()
                .when().get(ResourcePaths.DISHES + "/{id}", 1, 23)
                .then()
                .statusCode(200)
                .extract().asString();

        JsonApprovals.verifyJson(response);
    }

    @SuccessTest
    @DataSet(value = "/DishResourcesIT/dishes-scenario-2.yml")
    @ExpectedDataSet(value = "/DishResourcesIT/dishes-scenario-1.yml")
    @Transactional
    void shouldDeleteByIdDishesFromRestaurant() {

        given()
                .when().delete(ResourcePaths.DISHES + "/{id}", 1, 1)
                .then()
                .statusCode(204);

    }

    @SuccessTest
    @DataSet(value = "/DishResourcesIT/dishes-scenario-1.yml")
    @ExpectedDataSet(value = "/DishResourcesIT/dishes-scenario-2.yml")
    @Transactional
    void shouldCreateDishesFromRestaurant() {
        DishRequest request = new DishRequest();
        request.setName("X-Burguer");
        request.setPrice(new BigDecimal("9.99"));
        request.setCategory(Category.MAIN_COURSE);
        request.setIsAvailable(Boolean.TRUE);

        String response = given()
                .when().contentType("application/json").body(request)
                .post(ResourcePaths.DISHES, 1)
                .then()
                .statusCode(201)
                .extract().header("location");

        Approvals.verify(response);

    }

    @ErrorTest
    void shouldShowErrorOnCreateDishesFromRestaurantWithoutName() {

        String json = "{\"category\":\"MAIN_COURSE\",\"isAvailable\":true,\"price\":9.99}";

        String response = given()
                .when().contentType("application/json").body(json)
                .post(ResourcePaths.DISHES, 1)
                .then()
                .statusCode(400)
                .extract().response().body().asString();

        JsonApprovals.verifyJson(response);

    }

    @SuccessTest
    @DataSet(value = "/DishResourcesIT/dishes-scenario-2.yml")
    @ExpectedDataSet(value = "/DishResourcesIT/dishes-scenario-2-modified.yml")
    @Transactional
    void shouldUpdateNameDishesFromRestaurant() {
        DishRequest request = new DishRequest();
        request.setName("X-Burguer Duo");
        request.setPrice(new BigDecimal("9.99"));
        request.setCategory(Category.MAIN_COURSE);
        request.setIsAvailable(Boolean.TRUE);

        given()
                .when().contentType("application/json").body(request)
                .put(ResourcePaths.DISHES + "/{id}", 1, 1)
                .then()
                .statusCode(204);

    }

}
