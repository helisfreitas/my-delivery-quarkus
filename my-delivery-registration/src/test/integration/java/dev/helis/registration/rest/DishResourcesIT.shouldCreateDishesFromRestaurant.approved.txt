http://localhost:8081/restaurants/1/dishes/1